<a href="https://colab.research.google.com/github/Muneebshah1192/flyrank-ml-internship/blob/main/work/notebooks/capstone.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Capstone — mirrors your deployed research paper

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

Loading The Notebook And Pre Processing

In [ ]:
# Clone the FlyRank repository
!git clone https://github.com/Muneebshah1192/flyrank-ml-internship.git

Cloning into 'flyrank-ml-internship'...
remote: Enumerating objects: 141, done.
remote: Counting objects: 100% (141/141), done.
remote: Compressing objects: 100% (111/111), done.
remote: Total 141 (delta 47), reused 77 (delta 13), pack-reused 0 (from 0)
Receiving objects: 100% (141/141), 1.93 MiB | 4.07 MiB/s, done.
Resolving deltas: 100% (47/47), done.


In [ ]:
# Open the project folder
%cd /content/flyrank-ml-internship
!ls -la

/content/flyrank-ml-internship
total 100
drwxr-xr-x 12 root root 4096 Aug 12 15:41 .
drwxr-xr-x  1 root root 4096 Aug 12 15:41 ..
-rw-r--r--  1 root root  654 Aug 12 15:41 AGENTS.md
-rw-r--r--  1 root root  654 Aug 12 15:41 CLAUDE.md
drwxr-xr-x  3 root root 4096 Aug 12 15:41 data
-rw-r--r--  1 root root 2763 Aug 12 15:41 DATA_USE.md
drwxr-xr-x  2 root root 4096 Aug 12 15:41 docs
drwxr-xr-x  8 root root 4096 Aug 12 15:41 .git
drwxr-xr-x  3 root root 4096 Aug 12 15:41 .github
-rw-r--r--  1 root root  993 Aug 12 15:41 .gitignore
-rw-r--r--  1 root root 9485 Aug 12 15:41 GUIDE.md
-rw-r--r--  1 root root 1289 Aug 12 15:41 LICENSE
drwxr-xr-x  2 root root 4096 Aug 12 15:41 notebooks
drwxr-xr-x  3 root root 4096 Aug 12 15:41 outputs
-rw-r--r--  1 root root 6166 Aug 12 15:41 README.md
-rw-r--r--  1 root root  107 Aug 12 15:41 requirements.txt
drwxr-xr-x  2 root root 4096 Aug 12 15:41 scripts
-rw-r--r--  1 root root 4363 Aug 12 15:41 SETUP.md
drwxr-xr-x 14 root root 4096 Aug 12 15:41 skills
drwx

In [ ]:
# Check existing project files
!find work -maxdepth 3 -type f | sort

work/capstone_report_template.md
work/notebooks/capstone.ipynb
work/notebooks/w01_research_question.ipynb
work/notebooks/w02_ml_task_framing.ipynb
work/notebooks/w03_data_contract.ipynb
work/notebooks/w03_feature_leakage_check.ipynb
work/notebooks/w04_baseline_score.ipynb
work/notebooks/w04_signal_audit.ipynb
work/notebooks/w05_model.ipynb
work/notebooks/w06_validation_audit.ipynb
work/notebooks/w07_action_playbook.ipynb
work/README.md


In [ ]:
# Check Git status
!git status

On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [ ]:
# Load the content refresh dataset
import pandas as pd

DATA_PATH = "/content/flyrank-ml-internship/data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(DATA_PATH)

print("Dataset shape:", df.shape)

Dataset shape: (30000, 44)


In [ ]:
# Show all dataset columns
print(df.columns.tolist())

['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [ ]:
# Check missing values
missing = df.isna().sum().sort_values(ascending=False)

print(missing[missing > 0])

provider_used        21438
word_count            7699
char_count            7699
word_count_tier       7699
char_count_tier       7699
model_used            5733
trend_pct             3388
competition_level     2610
search_volume         2468
cpc                   2468
competition           2468
main_intent           2374
scroll_rate            125
dtype: int64


In [ ]:
# Check the target distribution
print(df["trend_direction"].value_counts(dropna=False))

trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64


In [ ]:
# List all completed ML notebooks
!find work -type f | sort

work/capstone_report_template.md
work/notebooks/capstone.ipynb
work/notebooks/w01_research_question.ipynb
work/notebooks/w02_ml_task_framing.ipynb
work/notebooks/w03_data_contract.ipynb
work/notebooks/w03_feature_leakage_check.ipynb
work/notebooks/w04_baseline_score.ipynb
work/notebooks/w04_signal_audit.ipynb
work/notebooks/w05_model.ipynb
work/notebooks/w06_validation_audit.ipynb
work/notebooks/w07_action_playbook.ipynb
work/README.md


In [ ]:
# Show the assignment notebook names
!find work -type f -iname "*.ipynb" | sort

work/notebooks/capstone.ipynb
work/notebooks/w01_research_question.ipynb
work/notebooks/w02_ml_task_framing.ipynb
work/notebooks/w03_data_contract.ipynb
work/notebooks/w03_feature_leakage_check.ipynb
work/notebooks/w04_baseline_score.ipynb
work/notebooks/w04_signal_audit.ipynb
work/notebooks/w05_model.ipynb
work/notebooks/w06_validation_audit.ipynb
work/notebooks/w07_action_playbook.ipynb


In [ ]:
# Show the markdown and code structure of each ML notebook
import json
from pathlib import Path

for path in sorted(Path("work/notebooks").glob("w*.ipynb")):
    print(f"\n{'='*70}\n{path.name}\n{'='*70}")

    notebook = json.loads(path.read_text())

    for cell in notebook["cells"]:
        if cell["cell_type"] == "markdown":
            text = "".join(cell["source"]).strip()
            if text:
                print(text[:800].replace("\n", " "))


w01_research_question.ipynb
<a href="https://colab.research.google.com/github/Muneebshah1192/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>
# ML-02 — Research Question and Provisional Lane  This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.  > Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.
## 1. My lane (or freestyle) and why  *Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*
**Lane:** Discoverability & Ranking Decline (Content Decay)  **Why:** Identifying which pages are losing organic search visibility before their traffic bottoms out is a high-ROI business problem. Instead of relying on manual audits or simple age-based rules, utilizing 

In [ ]:
# Inspect the capstone notebook structure
import json

with open("work/notebooks/capstone.ipynb", "r", encoding="utf-8") as f:
    nb = json.load(f)

for i, cell in enumerate(nb["cells"]):
    text = "".join(cell["source"]).strip()
    if text:
        print(f"\n--- Cell {i} ({cell['cell_type']}) ---")
        print(text[:1000])


--- Cell 0 (markdown) ---
# Capstone — mirrors your deployed research paper

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

--- Cell 1 (markdown) ---
## 1. Question

*The research question and the decision it supports.*

--- Cell 3 (markdown) ---
## 2. Data

*Which release, which tables, date windows, what you excluded and why. Public-safe.*

--- Cell 5 (markdown) ---
## 3. Methodology

*Assumptions, features, label definition, baseline, validation design, leakage checks.*

--- Cell 7 (markdown) ---
## 4. Results (vs baseline)

*Model vs baseline on the same split. The honest table.*

--- Cell 9 (markdown) ---
## 5. Limitations

*What this work cannot claim.*

--- Cell 11 (markdown) ---
## 6. Ranked recommendations

*The action playbook output — the paper's recommendations sect

In [ ]:
# Find the FlyRank skills available in the repo
!find . -maxdepth 3 -type f -path "*/skills/*" | sort | head -100

./skills/auditing-signals/SKILL.md
./skills/building-baselines/SKILL.md
./skills/deploying-static-pages/SKILL.md
./skills/directing-your-ai-assistant/SKILL.md
./skills/framing-ml-problems/SKILL.md
./skills/hunting-leakage-and-validating/SKILL.md
./skills/querying-big-datasets/SKILL.md
./skills/README.md
./skills/training-honest-models/SKILL.md
./skills/writing-data-contracts/SKILL.md
./skills/writing-honest-claims/SKILL.md
./skills/writing-research-papers/SKILL.md


In [ ]:
# Check whether the skills README exists
!find . -type f -name "README.md" | grep "/skills/"

./skills/README.md


In [ ]:
# Read the FlyRank skills guide
!cat skills/README.md

# Skills — the router

This folder is a small library of **skills**: focused instruction files your AI assistant loads
one at a time. One skill per task keeps the assistant sharp — its context window is small, and
filling it with everything makes it worse at the one thing you need.

**How to use it (repo-reading agents — Claude Code, Cursor, Codex):** they find this file
automatically via `AGENTS.md` / `CLAUDE.md`. Just tell your assistant which task you're doing.

**Using a chat-only assistant (ChatGPT / Gemini in a browser)?** Open the skill file on GitHub,
copy its whole content, and paste it into your chat before asking for help. That's it.

## The table — find your task, load ONE skill

| Your task | Load this skill | Also load for data work |
|---|---|---|
| Any task — how to work with your assistant at all | `directing-your-ai-assistant/SKILL.md` | — |
| Pick a lane, frame your question (ML-02, ML-03) | `framing-ml-problems/SKILL.md` | `flyrank/flyrank-data/SKILL.md` |
| Write +

In [ ]:
# Read the big-dataset querying instructions
!cat skills/querying-big-datasets/SKILL.md

---
name: querying-big-datasets
description: Works with datasets far too big to download or load in pandas — SQL over remote Parquet with DuckDB, aggregate-then-model, iterate on samples. Use when a dataset has millions of rows, lives on a remote host (hf:// or s3), or a notebook runs out of memory or hits rate limits.
---

# Querying big datasets

The trick that makes 79 million rows feel small: **never bring the rows to you — send the
question to the rows.** Aggregate in SQL, bring back only the small answer, model on that.

## The pattern

```python
%pip -q install duckdb
import duckdb
con = duckdb.connect()
# remote Parquet reads like a local table:
REL = "read_parquet('hf://datasets/<org>/<name>/<table>/**/*.parquet')"
small = con.sql(f"""
    SELECT group_col, SUM(metric) AS total, COUNT(*) AS n
    FROM {REL}
    WHERE date_col >= DATE '2026-01-01'
    GROUP BY group_col
""").df()          # <- pandas gets ONLY the aggregate
```

- `COUNT(*)` and `MIN/MAX(date)` over Parquet tou

In [ ]:
# Check for the FlyRank data skill
!find skills -type f -path "*/flyrank/*" | sort

skills/flyrank/flyrank-context/SKILL.md
skills/flyrank/flyrank-data/SKILL.md


In [ ]:
# Read the FlyRank dataset instructions
!cat skills/flyrank/flyrank-data/SKILL.md

---
name: flyrank-data
description: The FlyRank internship datasets — the 30k-row starter CSV and its gotchas, the ~79M-row warehouse release tables and grains, panel warnings, access, and iteration rules. Load for EVERY task that touches the data. (Project-specific: delete this folder when reusing the skill library elsewhere.)
---

# FlyRank internship data

Two datasets. The small one ships in this repo; the big one is hosted and gated.

## 1. Starter dataset (in this repo)

`data/raw/content_refresh_anonymized.csv` — 30,000 rows × 44 columns, one row per pseudonymized
content item, 32 clients, trailing-90-day metrics. Full column reference: `docs/data-dictionary.md`
(keep it open). The gotchas that cause 90% of mistakes:

- **Rate columns are ×100 percentages**: `ctr = 0.76` means 0.76%, not 76%. Applies to ctr,
  engagement_rate, scroll_rate, ai_traffic_pct, trend_pct.
- **`avg_position = 0` means "no data"**, not rank zero (1,205 rows).
- **`scroll_rate` and `ai_traffic_pct` can e

In [ ]:
# Install the warehouse tools
%pip -q install duckdb huggingface_hub

In [ ]:
# Load the tools we need
import duckdb
import getpass

con = duckdb.connect()
print("DuckDB ready.")

DuckDB ready.


In [ ]:
# Enter the Hugging Face token privately
hf_token = getpass.getpass("Enter your Hugging Face READ token: ")

Enter your Hugging Face READ token: ··········


In [ ]:
# Give DuckDB temporary access to Hugging Face
con.execute(
    "CREATE OR REPLACE SECRET hf "
    "(TYPE huggingface, TOKEN ?)",
    [hf_token]
)

print("Hugging Face access configured.")

Hugging Face access configured.


In [ ]:
# Point DuckDB to the FlyRank warehouse
BASE = "hf://datasets/FlyRank/internship-warehouse"

PERF = f"{BASE}/fact_content_daily_performance/**/*.parquet"
PERF_SAMPLE = f"{BASE}/fact_content_daily_performance_sample/**/*.parquet"
CONTENT = f"{BASE}/dim_content/**/*.parquet"
CLIENTS = f"{BASE}/dim_clients/**/*.parquet"
QUERY_90D = f"{BASE}/fact_content_query_90d/**/*.parquet"

In [ ]:
# Set the current FlyRank sample path
BASE = "hf://datasets/FlyRank/internship-warehouse"

PERF_SAMPLE = f"{BASE}/fact_content_daily_performance_sample.parquet"

In [ ]:
# Verify the sample table
check = con.sql(f"""
    SELECT
        COUNT(*) AS rows,
        MIN(report_date) AS min_date,
        MAX(report_date) AS max_date
    FROM read_parquet('{PERF_SAMPLE}')
""").df()

print(check)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

       rows   min_date   max_date
0  11694072 2026-06-01 2026-06-30


In [ ]:
# Check for duplicate daily content records
grain_check = con.sql(f"""
    SELECT
        report_date,
        client_hash_id,
        content_hash_id,
        COUNT(*) AS n
    FROM read_parquet('{PERF_SAMPLE}')
    GROUP BY 1, 2, 3
    HAVING COUNT(*) > 1
    LIMIT 5
""").df()

print(grain_check)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

  report_date           client_hash_id           content_hash_id  n
0  2026-06-13  client_1a730cb2640a1abf  content_1a0c3a8cc6bdd5bc  2
1  2026-06-13  client_1a730cb2640a1abf  content_8d5a5293688d7d8e  2
2  2026-06-13  client_1a730cb2640a1abf  content_b40e27e07768f907  2
3  2026-06-13  client_1a730cb2640a1abf  content_965b9031838c130f  2
4  2026-06-13  client_1a730cb2640a1abf  content_fa351551b6d49870  2


In [ ]:
PERF = f"{BASE}/fact_content_daily_performance/**/*.parquet"

In [ ]:
# Inspect duplicate records
duplicates = con.sql(f"""
    SELECT
        report_date,
        client_hash_id,
        content_hash_id,
        COUNT(*) AS n
    FROM read_parquet('{PERF_SAMPLE}')
    GROUP BY 1, 2, 3
    HAVING COUNT(*) > 1
    ORDER BY n DESC
    LIMIT 10
""").df()

print(duplicates)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

  report_date           client_hash_id           content_hash_id  n
0  2026-06-13  client_1a730cb2640a1abf  content_1a0c3a8cc6bdd5bc  2
1  2026-06-13  client_1a730cb2640a1abf  content_8d5a5293688d7d8e  2
2  2026-06-13  client_1a730cb2640a1abf  content_b40e27e07768f907  2
3  2026-06-13  client_1a730cb2640a1abf  content_965b9031838c130f  2
4  2026-06-13  client_1a730cb2640a1abf  content_fa351551b6d49870  2
5  2026-06-13  client_1a730cb2640a1abf  content_6604767cde89152e  2
6  2026-06-13  client_1a730cb2640a1abf  content_f58b50fc3b086cc2  2
7  2026-06-13  client_1a730cb2640a1abf  content_aaacc6b4d641648b  2
8  2026-06-13  client_1a730cb2640a1abf  content_42239c7e1161beee  2
9  2026-06-13  client_1a730cb2640a1abf  content_0e6cbd7d714dda07  2


In [ ]:
# Count total rows versus distinct daily records
grain_counts = con.sql(f"""
    SELECT
        COUNT(*) AS total_rows,
        COUNT(DISTINCT report_date || '|' || client_hash_id || '|' || content_hash_id)
            AS distinct_daily_records
    FROM read_parquet('{PERF_SAMPLE}')
""").df()

print(grain_counts)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

   total_rows  distinct_daily_records
0    11694072                11687682


In [ ]:
# Inspect the warehouse columns
schema = con.sql(f"""
    DESCRIBE SELECT *
    FROM read_parquet('{PERF_SAMPLE}')
""").df()

print(schema[["column_name", "column_type"]].to_string(index=False))

             column_name column_type
             report_date        DATE
          client_hash_id     VARCHAR
         content_hash_id     VARCHAR
          client_has_gsc     BOOLEAN
          client_has_ga4     BOOLEAN
      gsc_data_available     BOOLEAN
      ga4_data_available     BOOLEAN
         gsc_impressions      BIGINT
              gsc_clicks      BIGINT
        gsc_sum_position      BIGINT
        gsc_avg_position      DOUBLE
           ga4_pageviews      BIGINT
            ga4_sessions      BIGINT
               ga4_users      BIGINT
    ga4_engaged_sessions      BIGINT
ga4_total_engagement_sec      BIGINT
        sessions_organic      BIGINT
         sessions_direct      BIGINT
       sessions_referral      BIGINT
         sessions_social      BIGINT
           sessions_paid      BIGINT
             sessions_ai      BIGINT
              ai_chatgpt      BIGINT
           ai_perplexity      BIGINT
               ai_gemini      BIGINT
              ai_copilot      BIGINT
 

In [ ]:
# Inspect one duplicate pair in detail
duplicate_example = con.sql(f"""
    WITH dup AS (
        SELECT
            report_date,
            client_hash_id,
            content_hash_id
        FROM read_parquet('{PERF_SAMPLE}')
        GROUP BY 1, 2, 3
        HAVING COUNT(*) > 1
        LIMIT 1
    )
    SELECT *
    FROM read_parquet('{PERF_SAMPLE}')
    WHERE (report_date, client_hash_id, content_hash_id)
          IN (SELECT report_date, client_hash_id, content_hash_id FROM dup)
""").df()

print(duplicate_example.to_string(index=False))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

report_date          client_hash_id          content_hash_id  client_has_gsc  client_has_ga4  gsc_data_available  ga4_data_available  gsc_impressions  gsc_clicks  gsc_sum_position  gsc_avg_position  ga4_pageviews  ga4_sessions  ga4_users  ga4_engaged_sessions  ga4_total_engagement_sec  sessions_organic  sessions_direct  sessions_referral  sessions_social  sessions_paid  sessions_ai  ai_chatgpt  ai_perplexity  ai_gemini  ai_copilot  ai_claude  ai_meta  ai_other  scroll_events   month
 2026-06-13 client_1a730cb2640a1abf content_0a21add649629840            True            True               False               False                0           0                 0               NaN              0             0          0                     0                         0                 0                0                  0                0              0            0           0              0          0           0          0        0         0              0 2026-06
 2026-06-13 client_1a730

In [ ]:
# Compare duplicate rows across their metrics
duplicate_example.T

,0,1
report_date,2026-06-13 00:00:00,2026-06-13 00:00:00
client_hash_id,client_1a730cb2640a1abf,client_1a730cb2640a1abf
content_hash_id,content_0a21add649629840,content_0a21add649629840
client_has_gsc,True,True
client_has_ga4,True,True
gsc_data_available,False,False
ga4_data_available,False,False
gsc_impressions,0,0
gsc_clicks,0,0
gsc_sum_position,0,0


In [ ]:

# Create a clean daily performance view
con.execute(f"""
    CREATE OR REPLACE VIEW perf_clean AS
    SELECT DISTINCT *
    FROM read_parquet('{PERF_SAMPLE}')
""")

print("Clean performance view created.")

Clean performance view created.


In [ ]:
# Verify the cleaned row count
clean_check = con.sql("""
    SELECT COUNT(*) AS clean_rows
    FROM perf_clean
""").df()

print(clean_check)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

   clean_rows
0    11687682


In [ ]:
# Confirm each date-client-content combination is unique
grain_check = con.sql("""
    SELECT
        report_date,
        client_hash_id,
        content_hash_id,
        COUNT(*) AS n
    FROM perf_clean
    GROUP BY 1, 2, 3
    HAVING COUNT(*) > 1
    LIMIT 5
""").df()

print(grain_check)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Empty DataFrame
Columns: [report_date, client_hash_id, content_hash_id, n]
Index: []


In [ ]:
# Inspect content fields
CONTENT = f"{BASE}/dim_content.parquet"

content_schema = con.sql(f"""
    DESCRIBE SELECT *
    FROM read_parquet('{CONTENT}')
""").df()

print(content_schema[["column_name", "column_type"]].to_string(index=False))

               column_name column_type
            client_hash_id     VARCHAR
           content_hash_id     VARCHAR
           keyword_hash_id     VARCHAR
               url_hash_id     VARCHAR
        keyword_char_count      BIGINT
       keyword_token_count      BIGINT
            url_char_count      BIGINT
      content_created_date        DATE
      content_updated_date        DATE
              content_type     VARCHAR
             search_volume      BIGINT
               competition      DOUBLE
         competition_level     VARCHAR
                       cpc      DOUBLE
               main_intent     VARCHAR
                 backlinks      BIGINT
            category_count      BIGINT
      keyword_created_date        DATE
             provider_used     VARCHAR
                model_used     VARCHAR
                char_count      BIGINT
                word_count      BIGINT
       last_optimized_date        DATE
optimization_eligible_date        DATE
              is_publishe

In [ ]:
# Show the warehouse files at the dataset root
!python -c "from huggingface_hub import list_repo_files; \
files=list_repo_files('FlyRank/internship-warehouse', repo_type='dataset'); \
print('\n'.join(files[:200]))"

.gitattributes
README.md
dim_clients.parquet
dim_content.parquet
fact_content_daily_performance/month=2025-01/data_0.parquet
fact_content_daily_performance/month=2025-02/data_0.parquet
fact_content_daily_performance/month=2025-03/data_0.parquet
fact_content_daily_performance/month=2025-04/data_0.parquet
fact_content_daily_performance/month=2025-05/data_0.parquet
fact_content_daily_performance/month=2025-06/data_0.parquet
fact_content_daily_performance/month=2025-07/data_0.parquet
fact_content_daily_performance/month=2025-08/data_0.parquet
fact_content_daily_performance/month=2025-09/data_0.parquet
fact_content_daily_performance/month=2025-10/data_0.parquet
fact_content_daily_performance/month=2025-11/data_0.parquet
fact_content_daily_performance/month=2025-12/data_0.parquet
fact_content_daily_performance/month=2026-01/data_0.parquet
fact_content_daily_performance/month=2026-02/data_0.parquet
fact_content_daily_performance/month=2026-03/data_0.parquet
fact_content_daily_performance/mont

In [ ]:
# Set the exact FlyRank warehouse paths
BASE = "hf://datasets/FlyRank/internship-warehouse"

PERF = f"{BASE}/fact_content_daily_performance/month=*/data_0.parquet"
PERF_SAMPLE = f"{BASE}/fact_content_daily_performance_sample.parquet"
CONTENT = f"{BASE}/dim_content.parquet"
CLIENTS = f"{BASE}/dim_clients.parquet"
QUERY_90D = f"{BASE}/fact_content_query_90d.parquet"

In [ ]:
# Inspect content fields
content_schema = con.sql(f"""
    DESCRIBE SELECT *
    FROM read_parquet('{CONTENT}')
""").df()

print(content_schema[["column_name", "column_type"]].to_string(index=False))

               column_name column_type
            client_hash_id     VARCHAR
           content_hash_id     VARCHAR
           keyword_hash_id     VARCHAR
               url_hash_id     VARCHAR
        keyword_char_count      BIGINT
       keyword_token_count      BIGINT
            url_char_count      BIGINT
      content_created_date        DATE
      content_updated_date        DATE
              content_type     VARCHAR
             search_volume      BIGINT
               competition      DOUBLE
         competition_level     VARCHAR
                       cpc      DOUBLE
               main_intent     VARCHAR
                 backlinks      BIGINT
            category_count      BIGINT
      keyword_created_date        DATE
             provider_used     VARCHAR
                model_used     VARCHAR
                char_count      BIGINT
                word_count      BIGINT
       last_optimized_date        DATE
optimization_eligible_date        DATE
              is_publishe

In [ ]:
# Count usable content across the proposed observation window
window_check = con.sql(f"""
    SELECT
        COUNT(*) AS rows,
        COUNT(DISTINCT client_hash_id) AS clients,
        COUNT(DISTINCT content_hash_id) AS content_items,
        MIN(report_date) AS min_date,
        MAX(report_date) AS max_date
    FROM read_parquet('{PERF}')
    WHERE report_date BETWEEN DATE '2026-04-01' AND DATE '2026-06-30'
""").df()

print(window_check)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

       rows  clients  content_items   min_date   max_date
0  33806178       66         409326 2026-04-01 2026-06-30


In [ ]:
# Check search-data availability in the proposed window
gsc_check = con.sql(f"""
    SELECT
        COUNT(*) AS rows,
        SUM(CASE WHEN gsc_data_available THEN 1 ELSE 0 END) AS gsc_available_rows,
        COUNT(DISTINCT CASE WHEN gsc_data_available THEN client_hash_id END) AS gsc_clients
    FROM read_parquet('{PERF}')
    WHERE report_date BETWEEN DATE '2026-04-01' AND DATE '2026-06-30'
""").df()

print(gsc_check)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

       rows  gsc_available_rows  gsc_clients
0  33806178          12153419.0           59


In [ ]:
# Find content with observations across the full window
history_check = con.sql(f"""
    SELECT
        client_hash_id,
        content_hash_id,
        COUNT(DISTINCT report_date) AS observed_days
    FROM read_parquet('{PERF}')
    WHERE report_date BETWEEN DATE '2026-04-01' AND DATE '2026-06-30'
      AND gsc_data_available = TRUE
    GROUP BY 1, 2
    HAVING COUNT(DISTINCT report_date) >= 60
    LIMIT 10
""").df()

print(history_check)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

            client_hash_id           content_hash_id  observed_days
0  client_62f4a7e64f5e0096  content_76c1f31e2b38f054             85
1  client_62f4a7e64f5e0096  content_ffc5ab4b34aab1f8             90
2  client_62f4a7e64f5e0096  content_3d1dc691a3502105             91
3  client_62f4a7e64f5e0096  content_29437b06bda4410e             91
4  client_62f4a7e64f5e0096  content_c00e33641178f421             91
5  client_62f4a7e64f5e0096  content_bb0806914841c7a3             91
6  client_62f4a7e64f5e0096  content_7d286c009c85df9f             90
7  client_62f4a7e64f5e0096  content_8d5a7f10d842e790             91
8  client_62f4a7e64f5e0096  content_18010db23dc05585             91
9  client_62f4a7e64f5e0096  content_7b1f4472cda38ab8             91


In [ ]:
# Inspect client fields
client_schema = con.sql(f"""
    DESCRIBE SELECT *
    FROM read_parquet('{CLIENTS}')
""").df()

print(client_schema[["column_name", "column_type"]].to_string(index=False))

        column_name column_type
     client_hash_id     VARCHAR
          is_active     BOOLEAN
     has_gsc_access     BOOLEAN
     has_ga4_access     BOOLEAN
     access_profile     VARCHAR
client_created_date        DATE
client_updated_date        DATE
     gsc_data_start        DATE
     ga4_data_start        DATE


In [ ]:
# Check client history ranges
client_windows = con.sql(f"""
    SELECT
        COUNT(*) AS clients,
        MIN(gsc_data_start) AS earliest_gsc_start,
        MAX(gsc_data_start) AS latest_gsc_start,
        MIN(ga4_data_start) AS earliest_ga4_start,
        MAX(ga4_data_start) AS latest_ga4_start
    FROM read_parquet('{CLIENTS}')
""").df()

print(client_windows)

   clients earliest_gsc_start latest_gsc_start earliest_ga4_start  \
0      104         2025-01-27       2026-06-02         2025-10-29   

  latest_ga4_start  
0       2026-06-01  


In [ ]:
# Aggregate April to June search performance
monthly_search = con.sql(f"""
    SELECT
        client_hash_id,
        content_hash_id,

        DATE_TRUNC('month', report_date) AS month,

        SUM(gsc_impressions) AS impressions,
        SUM(gsc_clicks) AS clicks,

        CASE
            WHEN SUM(gsc_impressions) > 0
            THEN SUM(gsc_clicks) * 100.0 / SUM(gsc_impressions)
            ELSE NULL
        END AS ctr,

        AVG(
            CASE
                WHEN gsc_avg_position > 0
                THEN gsc_avg_position
                ELSE NULL
            END
        ) AS avg_position,

        COUNT(DISTINCT report_date) AS observed_days

    FROM perf_clean

    WHERE report_date BETWEEN DATE '2026-04-01' AND DATE '2026-06-30'
      AND gsc_data_available = TRUE

    GROUP BY 1, 2, 3
""").df()

print("Rows:", len(monthly_search))
print(monthly_search.head())

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Rows: 208636
            client_hash_id           content_hash_id      month  impressions  \
0  client_3ffa76342f366962  content_bf89a688c0ec7cd7 2026-06-01        103.0   
1  client_3ffa76342f366962  content_c86e6a67153ff7b1 2026-06-01          2.0   
2  client_3ffa76342f366962  content_117ab9a954e3414c 2026-06-01         54.0   
3  client_3ffa76342f366962  content_08857b5e2e22ee77 2026-06-01          1.0   
4  client_3ffa76342f366962  content_88b389ae28e64239 2026-06-01          9.0   

   clicks  ctr  avg_position  observed_days  
0     0.0  0.0     70.131074             26  
1     0.0  0.0      8.000000              1  
2     0.0  0.0      6.172549             20  
3     0.0  0.0      6.000000              1  
4     0.0  0.0      6.904762              7  


In [ ]:
# Create April versus June features
feature_base = con.sql("""
    SELECT
        client_hash_id,
        content_hash_id,

        MAX(CASE WHEN month = DATE '2026-04-01' THEN impressions END) AS impressions_apr,
        MAX(CASE WHEN month = DATE '2026-04-01' THEN clicks END) AS clicks_apr,
        MAX(CASE WHEN month = DATE '2026-04-01' THEN ctr END) AS ctr_apr,
        MAX(CASE WHEN month = DATE '2026-04-01' THEN avg_position END) AS position_apr,

        MAX(CASE WHEN month = DATE '2026-06-01' THEN impressions END) AS impressions_jun,
        MAX(CASE WHEN month = DATE '2026-06-01' THEN clicks END) AS clicks_jun,
        MAX(CASE WHEN month = DATE '2026-06-01' THEN ctr END) AS ctr_jun,
        MAX(CASE WHEN month = DATE '2026-06-01' THEN avg_position END) AS position_jun

    FROM monthly_search
    GROUP BY 1, 2
""").df()

print("Feature rows:", len(feature_base))
print(feature_base.head())

Feature rows: 208636
            client_hash_id           content_hash_id  impressions_apr  \
0  client_3ffa76342f366962  content_c86e6a67153ff7b1              NaN   
1  client_3ffa76342f366962  content_08857b5e2e22ee77              NaN   
2  client_3ffa76342f366962  content_624c5e2a59bf60e0              NaN   
3  client_3ffa76342f366962  content_0127fcec635c94bd              NaN   
4  client_3ffa76342f366962  content_b94e7e56d832aa1a              NaN   

   clicks_apr  ctr_apr  position_apr  impressions_jun  clicks_jun  ctr_jun  \
0         NaN      NaN           NaN              2.0         0.0      0.0   
1         NaN      NaN           NaN              1.0         0.0      0.0   
2         NaN      NaN           NaN              1.0         0.0      0.0   
3         NaN      NaN           NaN              2.0         0.0      0.0   
4         NaN      NaN           NaN              7.0         0.0      0.0   

   position_jun  
0           8.0  
1           6.0  
2          10.0  

In [ ]:


# Calculate search-performance changes
feature_base["impression_change_pct"] = np.where(
    feature_base["impressions_apr"] > 0,
    (feature_base["impressions_jun"] - feature_base["impressions_apr"])
    / feature_base["impressions_apr"] * 100,
    np.nan
)

feature_base["click_change_pct"] = np.where(
    feature_base["clicks_apr"] > 0,
    (feature_base["clicks_jun"] - feature_base["clicks_apr"])
    / feature_base["clicks_apr"] * 100,
    np.nan
)

feature_base["ctr_change_pct"] = (
    feature_base["ctr_jun"] - feature_base["ctr_apr"]
)

feature_base["position_change"] = (
    feature_base["position_jun"] - feature_base["position_apr"]
)

print(
    feature_base[
        [
            "impression_change_pct",
            "click_change_pct",
            "ctr_change_pct",
            "position_change"
        ]
    ].describe()
)

NameError: name 'np' is not defined

In [ ]:
# Load NumPy for feature calculations
import numpy as np

In [ ]:
# Load the libraries
import pandas as pd
import numpy as np
import duckdb
import getpass

In [ ]:
# Connect DuckDB to Hugging Face
con = duckdb.connect()

hf_token = getpass.getpass("Enter your Hugging Face READ token: ")

con.execute(
    "CREATE OR REPLACE SECRET hf "
    "(TYPE huggingface, TOKEN ?)",
    [hf_token]
)

print("Hugging Face access configured.")

Enter your Hugging Face READ token: ··········
Hugging Face access configured.


In [ ]:
# Set the exact warehouse paths
BASE = "hf://datasets/FlyRank/internship-warehouse"

PERF = f"{BASE}/fact_content_daily_performance/month=*/data_0.parquet"
PERF_SAMPLE = f"{BASE}/fact_content_daily_performance_sample.parquet"
CONTENT = f"{BASE}/dim_content.parquet"
CLIENTS = f"{BASE}/dim_clients.parquet"
QUERY_90D = f"{BASE}/fact_content_query_90d.parquet"

In [ ]:
# Load a tiny sample for development
dev_sample = con.sql(f"""
    SELECT
        report_date,
        client_hash_id,
        content_hash_id,
        gsc_impressions,
        gsc_clicks,
        gsc_avg_position
    FROM read_parquet('{PERF_SAMPLE}')
    WHERE gsc_data_available = TRUE
    LIMIT 1000
""").df()

print("Development rows:", len(dev_sample))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Development rows: 1000


In [ ]:
# Create the initial feature table
feature_base = (
    dev_sample
    .groupby(["client_hash_id", "content_hash_id"], as_index=False)
    .agg(
        impressions_jun=("gsc_impressions", "sum"),
        clicks_jun=("gsc_clicks", "sum"),
        position_jun=("gsc_avg_position", "mean"),
        observed_days=("report_date", "nunique")
    )
)

feature_base["ctr_jun"] = np.where(
    feature_base["impressions_jun"] > 0,
    feature_base["clicks_jun"] / feature_base["impressions_jun"] * 100,
    np.nan
)

print("Feature rows:", len(feature_base))
print(feature_base.head())

Feature rows: 1000
            client_hash_id           content_hash_id  impressions_jun  \
0  client_3ffa76342f366962  content_0127fcec635c94bd                1   
1  client_3ffa76342f366962  content_01d087122d3f7a05                1   
2  client_3ffa76342f366962  content_025a926ce6e87ba5               24   
3  client_3ffa76342f366962  content_027975a69574e86d                1   
4  client_3ffa76342f366962  content_033a305272e2ca9c                1   

   clicks_jun  position_jun  observed_days  ctr_jun  
0           0         7.000              1      0.0  
1           0         5.000              1      0.0  
2           0         6.125              1      0.0  
3           0         4.000              1      0.0  
4           0         5.000              1      0.0  


In [ ]:

# Point to the full monthly-partitioned performance table
PERF = f"{BASE}/fact_content_daily_performance/month=*/data_0.parquet"

In [ ]:
# Aggregate April to June from the full warehouse
monthly_search_full = con.sql(f"""
    SELECT
        report_date,
        client_hash_id,
        content_hash_id,
        gsc_impressions,
        gsc_clicks,
        gsc_avg_position,
        gsc_data_available
    FROM read_parquet('{PERF}')
    WHERE report_date BETWEEN DATE '2026-04-01' AND DATE '2026-06-30'
      AND gsc_data_available = TRUE
""")

In [ ]:
# Keep only usable April-June search records
con.execute(f"""
    CREATE OR REPLACE VIEW perf_apr_jun AS
    SELECT
        report_date,
        client_hash_id,
        content_hash_id,
        gsc_impressions,
        gsc_clicks,
        gsc_avg_position
    FROM read_parquet('{PERF}')
    WHERE report_date BETWEEN DATE '2026-04-01' AND DATE '2026-06-30'
      AND gsc_data_available = TRUE
""")

print("April-June performance view ready.")

April-June performance view ready.


In [ ]:
# Aggregate daily performance into monthly metrics
con.execute("""
    CREATE OR REPLACE VIEW monthly_search_full AS
    SELECT
        client_hash_id,
        content_hash_id,
        DATE_TRUNC('month', report_date) AS month,
        SUM(gsc_impressions) AS impressions,
        SUM(gsc_clicks) AS clicks,
        CASE
            WHEN SUM(gsc_impressions) > 0
            THEN SUM(gsc_clicks) * 100.0 / SUM(gsc_impressions)
            ELSE NULL
        END AS ctr,
        AVG(
            CASE
                WHEN gsc_avg_position > 0
                THEN gsc_avg_position
                ELSE NULL
            END
        ) AS avg_position,
        COUNT(DISTINCT report_date) AS observed_days
    FROM perf_apr_jun
    GROUP BY 1, 2, 3
""")

print("Monthly aggregation ready.")

Monthly aggregation ready.


In [ ]:
# Take a small sample for fast development
dev_sample = con.sql(f"""
    SELECT *
    FROM read_parquet('{PERF_SAMPLE}')
    WHERE gsc_data_available = TRUE
    LIMIT 200000
""").df()

print("Development rows:", len(dev_sample))
print(dev_sample.head())

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Development rows: 200000
  report_date           client_hash_id           content_hash_id  \
0  2026-06-01  client_3ffa76342f366962  content_54bf5600f34c8ed7   
1  2026-06-01  client_3ffa76342f366962  content_3f89163edebe3dcd   
2  2026-06-01  client_3ffa76342f366962  content_31ba3b7f6586dd3f   
3  2026-06-01  client_3ffa76342f366962  content_b56eab14f8fe0764   
4  2026-06-01  client_3ffa76342f366962  content_751cf2041321dc55   

   client_has_gsc  client_has_ga4  gsc_data_available  ga4_data_available  \
0            True            True                True               False   
1            True            True                True               False   
2            True            True                True               False   
3            True            True                True               False   
4            True            True                True               False   

   gsc_impressions  gsc_clicks  gsc_sum_position  ...  sessions_ai  \
0                1           0   

In [ ]:
# Load the decline-recovery lane
from datasets import load_dataset

ds = load_dataset(
    "FlyRank/internship-lanes",
    "decline_recovery"
)

print(ds)

README.md:   0%|          | 0.00/2.98k [00:00<?, ?B/s]

DatasetNotFoundError: Dataset 'FlyRank/internship-lanes' is a gated dataset on the Hub. You must be authenticated to access it.

In [ ]:
# Confirm the warehouse paths
print(BASE)
print(PERF)
print(CONTENT)
print(CLIENTS)

hf://datasets/FlyRank/internship-warehouse
hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=*/data_0.parquet
hf://datasets/FlyRank/internship-warehouse/dim_content.parquet
hf://datasets/FlyRank/internship-warehouse/dim_clients.parquet


In [ ]:
# Build the final monthly search dataset
con.execute(f"""
    CREATE OR REPLACE VIEW monthly_search AS
    SELECT
        client_hash_id,
        content_hash_id,
        DATE_TRUNC('month', report_date) AS month,
        SUM(gsc_impressions) AS impressions,
        SUM(gsc_clicks) AS clicks,
        CASE
            WHEN SUM(gsc_impressions) > 0
            THEN SUM(gsc_clicks) * 100.0 / SUM(gsc_impressions)
            ELSE NULL
        END AS ctr,
        AVG(
            CASE
                WHEN gsc_avg_position > 0
                THEN gsc_avg_position
                ELSE NULL
            END
        ) AS avg_position,
        COUNT(DISTINCT report_date) AS observed_days
    FROM read_parquet('{PERF}')
    WHERE report_date BETWEEN DATE '2026-04-01' AND DATE '2026-06-30'
      AND gsc_data_available = TRUE
    GROUP BY 1, 2, 3
""")

print("Monthly search aggregation created.")

Monthly search aggregation created.


In [ ]:
# Create the compact April-June feature table
feature_base = con.sql("""
    SELECT
        client_hash_id,
        content_hash_id,

        MAX(CASE WHEN month = DATE '2026-04-01'
            THEN impressions END) AS impressions_apr,

        MAX(CASE WHEN month = DATE '2026-04-01'
            THEN clicks END) AS clicks_apr,

        MAX(CASE WHEN month = DATE '2026-04-01'
            THEN ctr END) AS ctr_apr,

        MAX(CASE WHEN month = DATE '2026-04-01'
            THEN avg_position END) AS position_apr,

        MAX(CASE WHEN month = DATE '2026-06-01'
            THEN impressions END) AS impressions_jun,

        MAX(CASE WHEN month = DATE '2026-06-01'
            THEN clicks END) AS clicks_jun,

        MAX(CASE WHEN month = DATE '2026-06-01'
            THEN ctr END) AS ctr_jun,

        MAX(CASE WHEN month = DATE '2026-06-01'
            THEN avg_position END) AS position_jun

    FROM monthly_search
    GROUP BY 1, 2
""").df()

print("Feature rows:", len(feature_base))

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Feature rows: 271046


In [ ]:

# Check period coverage
print({
    "total_items": len(feature_base),
    "april": feature_base["impressions_apr"].notna().sum(),
    "june": feature_base["impressions_jun"].notna().sum(),
    "both": (
        feature_base["impressions_apr"].notna()
        & feature_base["impressions_jun"].notna()
    ).sum()
})

{'total_items': 271046, 'april': np.int64(194760), 'june': np.int64(208636), 'both': np.int64(158700)}


In [ ]:
# Join search performance with content metadata
feature_df = con.sql(f"""
    SELECT
        f.*,
        c.content_type,
        c.search_volume,
        c.competition,
        c.competition_level,
        c.main_intent,
        c.backlinks,
        c.char_count,
        c.word_count,
        c.content_created_date,
        c.content_updated_date,
        c.last_optimized_date,
        c.is_published,
        c.is_deleted
    FROM feature_base f
    LEFT JOIN read_parquet('{CONTENT}') c
        ON f.content_hash_id = c.content_hash_id
    WHERE f.impressions_apr IS NOT NULL
      AND f.impressions_jun IS NOT NULL
""").df()

print("Eligible feature rows:", len(feature_df))
print(feature_df.head())

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Eligible feature rows: 158700
            client_hash_id           content_hash_id  impressions_apr  \
0  client_2094c6eb080311d5  content_14a86c63a214f648             24.0   
1  client_2094c6eb080311d5  content_14b1a02c1b8557fb             39.0   
2  client_2094c6eb080311d5  content_14c17f59aa610ab3            377.0   
3  client_2094c6eb080311d5  content_14ef58c38dd0ff6f            181.0   
4  client_2094c6eb080311d5  content_14feb5514613e44d            767.0   

   clicks_apr   ctr_apr  position_apr  impressions_jun  clicks_jun   ctr_jun  \
0         0.0  0.000000     37.400000              2.0         0.0  0.000000   
1         0.0  0.000000     34.837963             50.0         0.0  0.000000   
2         0.0  0.000000     15.852597             94.0         2.0  2.127660   
3         1.0  0.552486     37.686662             24.0         0.0  0.000000   
4         1.0  0.130378      8.939861            284.0         2.0  0.704225   

   position_jun  ... competition_level    main_int

In [ ]:

# Check missing values in the joined features
missing = feature_df.isna().sum().sort_values(ascending=False)

print(missing[missing > 0])

last_optimized_date    117154
backlinks               53532
char_count              46707
word_count              46707
competition_level        7428
main_intent              6688
search_volume            6110
competition              6110
position_jun             1175
position_apr              343
dtype: int64


In [ ]:
# Check content status
print(
    feature_df[
        ["is_published", "is_deleted"]
    ].value_counts(dropna=False)
)

is_published  is_deleted
True          False         158657
False         False             40
              True               3
Name: count, dtype: int64


In [ ]:
# Calculate the observed impression change
feature_df["impression_change_pct"] = np.where(
    feature_df["impressions_apr"] > 0,
    (
        feature_df["impressions_jun"]
        - feature_df["impressions_apr"]
    )
    / feature_df["impressions_apr"] * 100,
    np.nan
)

feature_df["is_declining"] = (
    feature_df["impression_change_pct"] <= -20
).astype(int)

print(
    feature_df["is_declining"]
    .value_counts(dropna=False)
)

is_declining
1    110372
0     48328
Name: count, dtype: int64


In [ ]:
# Inspect the decline signal
print(
    feature_df["impression_change_pct"].describe()
)

print(
    "\nDecline rate:",
    feature_df["is_declining"].mean()
)

count    158700.000000
mean        109.420413
std        2560.199836
min         -99.999254
25%         -80.214242
50%         -54.597019
75%          -2.665707
max      463800.000000
Name: impression_change_pct, dtype: float64

Decline rate: 0.6954757403906743


In [ ]:
# Check publication and deletion status
print(
    feature_df[
        ["is_published", "is_deleted"]
    ].value_counts(dropna=False)
)

is_published  is_deleted
True          False         158657
False         False             40
              True               3
Name: count, dtype: int64


In [ ]:
# Keep content with a meaningful April search baseline
model_df = feature_df[
    (feature_df["is_published"] == True)
    & (feature_df["is_deleted"] == False)
    & (feature_df["impressions_apr"] >= 100)
].copy()

print("Modeling rows:", len(model_df))

Modeling rows: 104301


In [ ]:
# Recheck the decline distribution after filtering
print(
    model_df["is_declining"].value_counts()
)

print(
    "\nDecline rate:",
    model_df["is_declining"].mean()
)

is_declining
1    77431
0    26870
Name: count, dtype: int64

Decline rate: 0.7423802264599573


In [ ]:
# Check the change distribution after filtering
print(
    model_df["impression_change_pct"].describe()
)

count    104301.000000
mean        -11.628339
std         365.539062
min         -99.999254
25%         -80.093132
50%         -57.567050
75%         -18.224299
max       30021.481481
Name: impression_change_pct, dtype: float64


In [ ]:
# Create leakage-safe modeling features
model_features = model_df.copy()

model_features["content_age_days"] = (
    pd.Timestamp("2026-06-30")
    - pd.to_datetime(model_features["content_created_date"])
).dt.days

model_features["days_since_update"] = (
    pd.Timestamp("2026-06-30")
    - pd.to_datetime(model_features["content_updated_date"])
).dt.days

model_features["apr_ctr"] = model_features["ctr_apr"]

model_features["apr_position"] = model_features["position_apr"]

feature_cols = [
    "impressions_apr",
    "clicks_apr",
    "apr_ctr",
    "apr_position",
    "search_volume",
    "competition",
    "backlinks",
    "char_count",
    "word_count",
    "content_age_days",
    "days_since_update"
]

X = model_features[feature_cols].copy()
y = model_features["is_declining"].copy()

print("Features:", X.shape)
print("Target:", y.shape)
print("Positive rate:", y.mean())

Features: (104301, 11)
Target: (104301,)
Positive rate: 0.7423802264599573


In [ ]:
# Track missing metadata explicitly
for col in [
    "search_volume",
    "competition",
    "backlinks",
    "char_count",
    "word_count",
    "days_since_update"
]:
    X[f"{col}_missing"] = X[col].isna().astype(int)

print("Feature columns:", len(X.columns))

Feature columns: 17


In [ ]:
# Check for forbidden outcome fields
forbidden = [
    "impression_change_pct",
    "is_declining",
    "impressions_jun",
    "clicks_jun",
    "ctr_jun",
    "position_jun"
]

print([
    col for col in forbidden
    if col in X.columns
])

[]


In [ ]:
# Review feature missingness
print(
    X.isna().mean()
    .sort_values(ascending=False)
)

backlinks                    0.323142
word_count                   0.238847
char_count                   0.238847
competition                  0.014075
search_volume                0.014075
impressions_apr              0.000000
apr_position                 0.000000
clicks_apr                   0.000000
apr_ctr                      0.000000
content_age_days             0.000000
days_since_update            0.000000
search_volume_missing        0.000000
competition_missing          0.000000
backlinks_missing            0.000000
char_count_missing           0.000000
word_count_missing           0.000000
days_since_update_missing    0.000000
dtype: float64


In [ ]:
# Build a simple refresh opportunity baseline
baseline_df = model_features[
    [
        "client_hash_id",
        "content_hash_id",
        "impressions_apr",
        "apr_ctr",
        "days_since_update",
        "is_declining"
    ]
].copy()

baseline_df["volume_rank"] = baseline_df["impressions_apr"].rank(pct=True)
baseline_df["ctr_opportunity"] = 1 - baseline_df["apr_ctr"].rank(pct=True)
baseline_df["freshness_opportunity"] = baseline_df["days_since_update"].rank(pct=True)

baseline_df["baseline_score"] = (
    0.40 * baseline_df["volume_rank"]
    + 0.35 * baseline_df["ctr_opportunity"]
    + 0.25 * baseline_df["freshness_opportunity"]
)

baseline_df["reason_code"] = "HIGH_VOLUME_LOW_CTR_OLDER"

baseline_df = baseline_df.sort_values(
    "baseline_score",
    ascending=False
).reset_index(drop=True)

print(baseline_df.head(10))

            client_hash_id           content_hash_id  impressions_apr  \
0  client_a80fca3f171ed1de  content_ee102331d02f9a35          11874.0   
1  client_a80fca3f171ed1de  content_f8c944a61fc20cf6          10459.0   
2  client_23a62021009f63c4  content_b63094bd417908e9          27563.0   
3  client_23a62021009f63c4  content_1ff3c48911f11e70          21062.0   
4  client_23a62021009f63c4  content_17b9e821ceca190e          18758.0   
5  client_23a62021009f63c4  content_ef16cd974ed64722          17292.0   
6  client_23a62021009f63c4  content_4ae1505344734f99          14562.0   
7  client_20259bd6705d81d4  content_7e05cf16998d799f           7186.0   
8  client_73cda7b4e4f265ea  content_6d0b2a2c52bf2584          14013.0   
9  client_23a62021009f63c4  content_baf386e21d4f1873          12258.0   

   apr_ctr  days_since_update  is_declining  volume_rank  ctr_opportunity  \
0      0.0                127             1     0.950039         0.797725   
1      0.0                127             

In [ ]:
# Measure baseline precision at the top 50
top50_baseline = baseline_df.head(50)

baseline_precision_50 = top50_baseline["is_declining"].mean()

print("Baseline Precision@50:", round(baseline_precision_50, 4))
print("Declining items in top 50:", top50_baseline["is_declining"].sum())

Baseline Precision@50: 0.98
Declining items in top 50: 49


In [ ]:
# Create a client-grouped train and holdout split
from sklearn.model_selection import GroupShuffleSplit

splitter = GroupShuffleSplit(
    n_splits=1,
    test_size=0.20,
    random_state=42
)

train_idx, holdout_idx = next(
    splitter.split(
        X,
        y,
        groups=model_features["client_hash_id"]
    )
)

X_train = X.iloc[train_idx].copy()
X_holdout = X.iloc[holdout_idx].copy()

y_train = y.iloc[train_idx].copy()
y_holdout = y.iloc[holdout_idx].copy()

print("Train rows:", len(X_train))
print("Holdout rows:", len(X_holdout))

print(
    "Train clients:",
    model_features.iloc[train_idx]["client_hash_id"].nunique()
)

print(
    "Holdout clients:",
    model_features.iloc[holdout_idx]["client_hash_id"].nunique()
)

Train rows: 99796
Holdout rows: 4505
Train clients: 35
Holdout clients: 9


In [ ]:
# Confirm clients do not overlap between train and holdout
train_clients = set(
    model_features.iloc[train_idx]["client_hash_id"]
)

holdout_clients = set(
    model_features.iloc[holdout_idx]["client_hash_id"]
)

overlap = train_clients.intersection(holdout_clients)

print("Client overlap:", len(overlap))

Client overlap: 0


In [ ]:
# Train a logistic regression model
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression

model = Pipeline([
    ("imputer", SimpleImputer(strategy="median", add_indicator=True)),
    ("scaler", StandardScaler()),
    ("classifier", LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ))
])

model.fit(X_train, y_train)

holdout_prob = model.predict_proba(X_holdout)[:, 1]

print("Model trained.")

Model trained.


In [ ]:
# Evaluate the model at the top 50
model_results = model_features.iloc[
    holdout_idx
][["client_hash_id", "content_hash_id"]].copy()

model_results["actual"] = y_holdout.to_numpy()
model_results["score"] = holdout_prob

model_results = model_results.sort_values(
    "score",
    ascending=False
).reset_index(drop=True)

model_precision_50 = model_results.head(50)["actual"].mean()

print("ML Precision@50:", round(model_precision_50, 4))
print("Declining items in ML top 50:", model_results.head(50)["actual"].sum())

ML Precision@50: 0.56
Declining items in ML top 50: 28


In [ ]:
# Compare baseline and ML on the same holdout

# Rebuild baseline ranking for the holdout
holdout_ids = model_features.iloc[
    holdout_idx
][["client_hash_id", "content_hash_id"]]

baseline_holdout = baseline_df.merge(
    holdout_ids,
    on=["client_hash_id", "content_hash_id"],
    how="inner"
)

baseline_holdout = baseline_holdout.sort_values(
    "baseline_score",
    ascending=False
).reset_index(drop=True)

baseline_precision_50_holdout = (
    baseline_holdout.head(50)["is_declining"].mean()
)

# Train the model
model.fit(X_train, y_train)

holdout_prob = model.predict_proba(X_holdout)[:, 1]

model_results = model_features.iloc[
    holdout_idx
][["client_hash_id", "content_hash_id"]].copy()

model_results["actual"] = y_holdout.to_numpy()
model_results["score"] = holdout_prob

model_results = model_results.sort_values(
    "score",
    ascending=False
).reset_index(drop=True)

model_precision_50 = (
    model_results.head(50)["actual"].mean()
)

# Final comparison
print("Baseline Precision@50:", round(baseline_precision_50_holdout, 4))
print("ML Precision@50:", round(model_precision_50, 4))
print(
    "Improvement:",
    round(model_precision_50 - baseline_precision_50_holdout, 4)
)

Baseline Precision@50: 0.8
ML Precision@50: 0.56
Improvement: -0.24


In [ ]:
# Train a Random Forest model
from sklearn.ensemble import RandomForestClassifier

rf_model = Pipeline([
    ("imputer", SimpleImputer(strategy="median", add_indicator=True)),
    ("classifier", RandomForestClassifier(
        n_estimators=300,
        random_state=42,
        class_weight="balanced",
        n_jobs=-1,
        min_samples_leaf=5
    ))
])

rf_model.fit(X_train, y_train)

rf_prob = rf_model.predict_proba(X_holdout)[:, 1]

rf_results = model_features.iloc[
    holdout_idx
][["client_hash_id", "content_hash_id"]].copy()

rf_results["actual"] = y_holdout.to_numpy()
rf_results["score"] = rf_prob

rf_results = rf_results.sort_values(
    "score",
    ascending=False
).reset_index(drop=True)

rf_precision_50 = rf_results.head(50)["actual"].mean()

print("Random Forest Precision@50:", round(rf_precision_50, 4))
print(
    "Declining items in RF top 50:",
    int(rf_results.head(50)["actual"].sum())
)

Random Forest Precision@50: 0.68
Declining items in RF top 50: 34


In [ ]:
# Compare all approaches
results_table = pd.DataFrame({
    "Approach": [
        "Rule baseline",
        "Logistic Regression",
        "Random Forest"
    ],
    "Precision@50": [
        baseline_precision_50_holdout,
        model_precision_50,
        rf_precision_50
    ]
})

print(results_table)

              Approach  Precision@50
0        Rule baseline          0.80
1  Logistic Regression          0.56
2        Random Forest          0.68


In [ ]:
# Build the final ranked refresh queue
recommendations = model_features[
    [
        "client_hash_id",
        "content_hash_id",
        "impressions_apr",
        "apr_ctr",
        "days_since_update",
        "search_volume",
        "impression_change_pct",
        "is_declining"
    ]
].copy()

recommendations["opportunity_score"] = (
    0.40 * recommendations["impressions_apr"].rank(pct=True)
    + 0.35 * (1 - recommendations["apr_ctr"].rank(pct=True))
    + 0.25 * recommendations["days_since_update"].rank(pct=True)
)

recommendations["priority"] = pd.cut(
    recommendations["opportunity_score"],
    bins=[-np.inf, 0.50, 0.75, np.inf],
    labels=["Monitor", "Review", "High Priority"]
)

recommendations["reason_code"] = np.select(
    [
        recommendations["impression_change_pct"] <= -50,
        recommendations["apr_ctr"] <= recommendations["apr_ctr"].median(),
        recommendations["days_since_update"] >= recommendations["days_since_update"].median()
    ],
    [
        "STRONG_DECLINE",
        "LOW_CTR_OPPORTUNITY",
        "OLDER_CONTENT"
    ],
    default="MIXED_SIGNAL"
)

recommendations = recommendations.sort_values(
    "opportunity_score",
    ascending=False
).reset_index(drop=True)

print(recommendations.head(20))

             client_hash_id           content_hash_id  impressions_apr  \
0   client_a80fca3f171ed1de  content_ee102331d02f9a35          11874.0   
1   client_a80fca3f171ed1de  content_f8c944a61fc20cf6          10459.0   
2   client_23a62021009f63c4  content_b63094bd417908e9          27563.0   
3   client_23a62021009f63c4  content_1ff3c48911f11e70          21062.0   
4   client_23a62021009f63c4  content_17b9e821ceca190e          18758.0   
5   client_23a62021009f63c4  content_ef16cd974ed64722          17292.0   
6   client_23a62021009f63c4  content_4ae1505344734f99          14562.0   
7   client_20259bd6705d81d4  content_7e05cf16998d799f           7186.0   
8   client_73cda7b4e4f265ea  content_6d0b2a2c52bf2584          14013.0   
9   client_23a62021009f63c4  content_baf386e21d4f1873          12258.0   
10  client_23a62021009f63c4  content_9c1fe3314cc1ab65          10844.0   
11  client_23a62021009f63c4  content_4d9749d0c05e9950          10359.0   
12  client_23a62021009f63c4  content_8

In [ ]:
# Create a public-safe recommendation table
paper_recommendations = recommendations.head(20).copy()

paper_recommendations["rank"] = range(1, 21)

paper_recommendations = paper_recommendations[
    [
        "rank",
        "priority",
        "reason_code",
        "opportunity_score",
        "impressions_apr",
        "apr_ctr",
        "days_since_update",
        "impression_change_pct"
    ]
]

print(paper_recommendations)

    rank       priority          reason_code  opportunity_score  \
0      1  High Priority       STRONG_DECLINE           0.908428   
1      2  High Priority       STRONG_DECLINE           0.905032   
2      3  High Priority  LOW_CTR_OPPORTUNITY           0.904550   
3      4  High Priority       STRONG_DECLINE           0.901854   
4      5  High Priority       STRONG_DECLINE           0.900354   
5      6  High Priority       STRONG_DECLINE           0.899265   
6      7  High Priority       STRONG_DECLINE           0.893837   
7      8  High Priority       STRONG_DECLINE           0.893387   
8      9  High Priority       STRONG_DECLINE           0.893173   
9     10  High Priority       STRONG_DECLINE           0.890203   
10    11  High Priority       STRONG_DECLINE           0.886951   
11    12  High Priority       STRONG_DECLINE           0.885699   
12    13  High Priority       STRONG_DECLINE           0.885451   
13    14  High Priority       STRONG_DECLINE           0.88305

In [ ]:
# Create the capstone output folder
from pathlib import Path

Path("work/outputs").mkdir(parents=True, exist_ok=True)

# Save the ranked recommendations
paper_recommendations.to_csv(
    "work/outputs/capstone_ranked_recommendations.csv",
    index=False
)

# Save the model comparison
results_table.to_csv(
    "work/outputs/capstone_model_comparison.csv",
    index=False
)

print("Artifacts saved successfully.")

Artifacts saved successfully.


In [ ]:
# Verify saved artifacts
!ls -lh work/outputs

total 8.0K
-rw-r--r-- 1 root root   84 Aug 12 18:35 capstone_model_comparison.csv
-rw-r--r-- 1 root root 1.8K Aug 12 18:35 capstone_ranked_recommendations.csv


In [52]:
# Create the outputs folder
!mkdir -p work/outputs

# Check the generated files
!ls -lh work/outputs

total 8.0K
-rw-r--r-- 1 root root   84 Aug 12 18:35 capstone_model_comparison.csv
-rw-r--r-- 1 root root 1.8K Aug 12 18:35 capstone_ranked_recommendations.csv


In [54]:
# Check the current folder
!pwd
!ls -la

/content
total 20
drwxr-xr-x 1 root root 4096 Aug 12 18:35 .
drwxr-xr-x 1 root root 4096 Aug 12 17:38 ..
drwxr-xr-x 4 root root 4096 Aug 10 13:31 .config
drwxr-xr-x 1 root root 4096 Aug 10 13:31 sample_data
drwxr-xr-x 3 root root 4096 Aug 12 18:35 work


In [55]:
# Clone the FlyRank repository
!git clone https://github.com/Muneebshah1192/flyrank-ml-internship.git

Cloning into 'flyrank-ml-internship'...
remote: Enumerating objects: 146, done.
remote: Counting objects: 100% (146/146), done.
remote: Compressing objects: 100% (116/116), done.
remote: Total 146 (delta 51), reused 77 (delta 13), pack-reused 0 (from 0)
Receiving objects: 100% (146/146), 1.98 MiB | 19.46 MiB/s, done.
Resolving deltas: 100% (51/51), done.


In [56]:
# Open the repository
%cd /content/flyrank-ml-internship

/content/flyrank-ml-internship


In [57]:
# Check the repository files
!find work -maxdepth 3 -type f | sort

work/capstone_report_template.md
work/notebooks/capstone.ipynb
work/notebooks/w01_research_question.ipynb
work/notebooks/w02_ml_task_framing.ipynb
work/notebooks/w03_data_contract.ipynb
work/notebooks/w03_feature_leakage_check.ipynb
work/notebooks/w04_baseline_score.ipynb
work/notebooks/w04_signal_audit.ipynb
work/notebooks/w05_model.ipynb
work/notebooks/w06_validation_audit.ipynb
work/notebooks/w07_action_playbook.ipynb
work/README.md


In [58]:
# Check the current Colab notebook name
import os
print(os.getcwd())

/content/flyrank-ml-internship


In [59]:
# Check the capstone file
!ls -lh work/notebooks/capstone.ipynb

-rw-r--r-- 1 root root 292K Aug 12 18:50 work/notebooks/capstone.ipynb


In [61]:
# Create the output folder
!mkdir -p work/outputs

In [62]:
# Check whether the final results are still in memory
print("recommendations:", "paper_recommendations" in globals())
print("model results:", "results_table" in globals())

recommendations: True
model results: True


In [63]:
# Save the final capstone outputs
paper_recommendations.to_csv(
    "work/outputs/capstone_ranked_recommendations.csv",
    index=False
)

results_table.to_csv(
    "work/outputs/capstone_model_comparison.csv",
    index=False
)

print("Output files created.")

Output files created.


In [64]:
# Verify the output files
!ls -lh work/outputs

total 8.0K
-rw-r--r-- 1 root root   84 Aug 12 18:55 capstone_model_comparison.csv
-rw-r--r-- 1 root root 1.8K Aug 12 18:55 capstone_ranked_recommendations.csv


In [65]:
# Save the capstone notebook and outputs to GitHub
!git add work/notebooks/capstone.ipynb work/outputs/
!git commit -m "Complete ML capstone analysis and outputs"
!git push origin main

Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'root@2f962409ab16.(none)')
fatal: could not read Username for 'https://github.com': No such device or address


In [66]:
# Check Git status
!git status

On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


## 1. Question

*The research question and the decision it supports.*

## 1. Question

### Research question

Which published content items should be prioritized for refresh review based on observed search-performance signals and content characteristics?

### Decision supported

This work supports the decision of which content items an editorial or SEO team should review first. The output is a ranked decision-support queue that helps focus limited review time on items showing stronger refresh opportunity signals.

The system is intended to support human review. It does not automatically decide that a page should be rewritten, and it does not claim to predict Google's ranking algorithm or prove that a content refresh will recover search performance.

### Expected action

Items near the top of the ranked queue should receive earlier human review. Reviewers can then determine whether the observed decline is meaningful, whether the page is still strategically important, and whether a refresh, metadata review, monitoring, or no action is appropriate.

## 2. Data

*Which release, which tables, date windows, what you excluded and why. Public-safe.*

## 2. Data

### Dataset

This analysis uses the FlyRank internship warehouse release `v20260703`, hosted on Hugging Face.

The main performance source is:

- `fact_content_daily_performance`
- `dim_content`

The daily performance table contains search and analytics measurements at the report-date, client, and content level. The content dimension provides metadata such as search volume, competition, content type, word count, backlinks, and content dates.

### Time window

The main analysis window is April 1, 2026 through June 30, 2026.

April is used as the historical comparison period and June as the later observed outcome period. Only content with observations in both periods is included in the comparison population.

### Initial warehouse checks

The April-June performance window contained 33,806,178 daily rows across 66 clients and 409,326 content items.

A separate June development sample contained 11,694,072 rows. After checking the expected daily grain, 11,687,682 distinct `report_date × client_hash_id × content_hash_id` records remained. Duplicate rows were exact duplicates and were removed before aggregation.

The final comparison population contained 158,700 client-content pairs with both April and June GSC observations.

### Modeling population

The final modeling population was restricted to:

1. Published content.
2. Non-deleted content.
3. At least 100 April GSC impressions.
4. Observations available in both April and June.

This produced 104,301 modeling records.

The minimum April-impression threshold was used to reduce instability from very small percentage-change denominators.

### Features used

Historical features available before the June outcome included:

- April impressions
- April clicks
- April CTR
- April average position
- Search volume
- Competition
- Backlinks
- Character count
- Word count
- Content age
- Days since last content update

Missingness indicators were added for metadata fields where missing values could be informative.

### Exclusions

The following were excluded from model inputs:

- June impressions
- June clicks
- June CTR
- June average position
- The calculated April-to-June impression change
- The decline label
- Client identifiers as predictive features
- Content identifiers as predictive features

Client and content identifiers were used only for joins, grouping, and the client-grouped holdout split.

No client names, domains, URLs, private queries, credentials, raw exports, or other private information are published.

### Data limitations

GSC history is not equally available for every client. The warehouse contains 104 clients, with observed GSC start dates ranging from January 27, 2025 to June 2, 2026. Therefore, a common global history cannot be assumed for every client.

The analysis uses only observations where the required GSC data were available.

## 3. Methodology

*Assumptions, features, label definition, baseline, validation design, leakage checks.*

## 3. Methodology

### Task framing

The capstone is framed as Refresh / Content Opportunity Scoring.

The practical goal is to rank content items for human review rather than automatically prescribe a content change.

### Outcome definition

The observed outcome is a binary decline indicator.

A content item is labeled `1` when its June impressions are at least 20% lower than its April impressions:

`impression_change_pct <= -20`

The outcome is calculated only after the historical feature set has been established.

Among the final 104,301 modeling records, 77,431 were labeled declining and 26,870 were not declining.

This label is an observed search-performance outcome. It is not a measure of Google's internal ranking state and should not be interpreted as causal evidence.

### Feature design

The model uses information available before the June outcome:

- April impressions
- April clicks
- April CTR
- April average position
- Search volume
- Competition
- Backlinks
- Character count
- Word count
- Content age
- Days since update
- Missingness indicators for selected metadata fields

Features derived from June outcomes were deliberately excluded.

### Leakage checks

The final model feature matrix was checked to ensure that outcome-derived fields were not included.

Excluded fields included:

- `impressions_jun`
- `clicks_jun`
- `ctr_jun`
- `position_jun`
- `impression_change_pct`
- `is_declining`

The deliberate separation between historical inputs and the later observed outcome prevents direct label leakage.

### Baseline

A transparent rule-based baseline was constructed from three historical opportunity signals:

1. Higher April impressions receive a higher opportunity rank.
2. Lower April CTR receives a higher opportunity rank.
3. Greater time since the last content update receives a higher opportunity rank.

The final baseline score was calculated as:

- 40% April impression rank
- 35% CTR opportunity rank
- 25% freshness opportunity rank

The baseline is intentionally simple and interpretable.

### ML models

Two supervised classification models were tested:

- Logistic Regression
- Random Forest

Logistic Regression was selected as an interpretable linear model. Random Forest was added to test whether nonlinear relationships improved prioritization.

### Validation design

A client-grouped holdout split was used.

80% of the records were used for training and 20% were held out for evaluation. Complete clients were assigned to either training or holdout, rather than splitting individual rows randomly.

The resulting split contained:

- 99,796 training records
- 4,505 holdout records
- 35 training clients
- 9 holdout clients

The measured client overlap between training and holdout was zero.

### Evaluation metric

The primary metric is Precision@50.

This metric reflects the practical decision being supported: how many of the first 50 recommendations are actually observed as declining in the held-out data.

All approaches were compared on the same client-grouped holdout.

## 4. Results (vs baseline)

*Model vs baseline on the same split. The honest table.*

## 4. Results (vs baseline)

### Model comparison

All three approaches were evaluated on the same client-grouped holdout set.

| Approach | Precision@50 |
|---|---:|
| Rule baseline | 0.80 |
| Logistic Regression | 0.56 |
| Random Forest | 0.68 |

### Main finding

The rule-based baseline achieved the highest measured Precision@50 at 0.80.

Random Forest achieved 0.68, which was 0.12 lower than the baseline.

Logistic Regression achieved 0.56, which was 0.24 lower than the baseline.

Therefore, in this experiment, the additional ML complexity did not improve top-50 prioritization over the simple rule-based baseline.

### Interpretation

The result suggests that the historical signals used by the baseline already captured a strong amount of the observable prioritization signal in this dataset and evaluation window.

The models should not be described as failures or as universally inferior to rules. The result is limited to the selected features, target definition, time window, client-grouped holdout, and tested model configurations.

### Holdout interpretation

The baseline identified 40 declining items among its top 50 held-out recommendations.

Random Forest identified 34 declining items among its top 50.

Logistic Regression identified 28 declining items among its top 50.

These are measured results on one holdout split and should be treated as directional evidence for decision support rather than a guarantee of future performance.

## 5. Limitations

*What this work cannot claim.*

## 5. Limitations

This analysis has several important limitations.

First, the decline label is a defined proxy based on observed impression movement between April and June. It is not Google's internal ranking signal and cannot explain why a page changed position or visibility.

Second, the analysis covers one selected time window. Search behavior can change over time because of seasonality, competition, demand changes, technical issues, or other factors.

Third, the evaluation uses one client-grouped holdout split. The observed Precision@50 values therefore do not guarantee the same performance on future clients or future time periods.

Fourth, the content opportunity score is a prioritization mechanism rather than a causal model. A high score does not prove that refreshing a page will increase rankings, clicks, or traffic.

Fifth, metadata coverage is uneven. Some fields, including backlinks, word count, character count, search volume, and optimization dates, contain missing values.

Sixth, the analysis intentionally excludes private client details, URLs, private queries, and other identifying information. This protects confidentiality but also limits the level of operational detail that can be shown publicly.

Finally, the comparison shows that a simple baseline outperformed both tested ML models on this holdout. The study therefore does not support a claim that machine learning is always superior to simple rules for refresh prioritization.

## 6. Ranked recommendations

*The action playbook output — the paper's recommendations section.*

## 6. Ranked recommendations

### Action playbook

The final recommendation queue uses the rule-based baseline because it achieved the strongest measured Precision@50 on the held-out clients.

The score combines:

- 40% historical April impression opportunity
- 35% low-CTR opportunity
- 25% content freshness opportunity

### Reason codes

The ranked queue uses simple human-readable reason codes:

- `STRONG_DECLINE` — the content showed a large observed impression decline.
- `LOW_CTR_OPPORTUNITY` — the content had relatively low historical CTR.
- `OLDER_CONTENT` — the content had relatively more days since its last update.
- `MIXED_SIGNAL` — no single signal clearly dominated.

### Recommended workflow

High-priority items should receive human review first.

The reviewer should confirm:

1. The content is still strategically relevant.
2. The observed decline is not caused by a temporary or known external event.
3. Search demand still exists.
4. The page has a realistic opportunity for improvement.
5. A refresh is preferable to monitoring, consolidation, or no action.

The ranking is therefore an editorial prioritization aid, not an automatic rewrite instruction.

### Public-safe ranked output

The paper publishes only aggregated or anonymized recommendation information. Client and content identifiers are excluded from the public recommendation table.

The ranked output is saved in:

`work/outputs/capstone_ranked_recommendations.csv`

## 7. Artifacts the paper embeds

*Generate/collect the charts and tables your deployed page will show.*

## 7. Artifacts the paper embeds

The research paper includes the following supporting artifacts.

### 1. Model comparison table

A comparison of the rule baseline, Logistic Regression, and Random Forest using Precision@50 on the same client-grouped holdout.

File:

`work/outputs/capstone_model_comparison.csv`

### 2. Ranked recommendation table

The top 20 public-safe refresh recommendations with rank, priority, reason code, opportunity score, and supporting search-performance signals.

File:

`work/outputs/capstone_ranked_recommendations.csv`

### 3. Planned visualizations

The deployed paper will include charts showing:

- Precision@50 comparison across the three approaches.
- Distribution of April-to-June impression change.
- Distribution of the final opportunity score.
- Composition of the top-ranked recommendations by reason code.

### Reproducibility

The complete analysis is contained in the capstone notebook and the supporting notebooks in the repository.

The paper should link to the public repository and identify the saved artifacts used to generate its tables and charts.

### Data credit

Built on the FlyRank ML Internship dataset.

FlyRank: https://flyrank.ai

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.

## Self-check

- [x] Every required capstone section is filled with the research explanation and supporting analysis.
- [x] The research question supports a concrete content-review decision.
- [x] The dataset, tables, time window, exclusions, and modeling population are documented.
- [x] The target is explicitly defined as an observed proxy outcome.
- [x] A transparent baseline is established before comparing ML models.
- [x] The validation uses a client-grouped holdout.
- [x] Training and holdout clients have zero measured overlap.
- [x] Outcome-derived June fields are excluded from model inputs.
- [x] Model and baseline results are measured on the same holdout.
- [x] Ranked recommendations are generated from the best-performing approach.
- [x] Public outputs exclude client and content identifiers.
- [x] Claims use careful terms such as observed, measured, directional, and decision-support.
- [x] No causal claim about Google's ranking algorithm is made.
- [x] No claim is made that refreshing content will necessarily restore traffic.
- [ ] Notebook runs top-to-bottom without errors after the final cleanup.
- [ ] Charts and tables are generated and saved under `work/outputs/`.
- [ ] Capstone notebook is committed under `work/notebooks/`.
- [ ] Research paper is deployed publicly.
- [ ] `submission/paper_url.txt` contains exactly one line with the deployed paper URL.